atc/db/component.go

```diff
@@ -68,28 +85,57 @@ func (c *component) Reload() (bool, error) {
 // component to across ATCs more evenly.
 func (c *component) IntervalElapsed() bool {
 	interval := c.interval
-	drift := time.Duration(c.rander.Int())%(2*time.Second) - time.Second
+	drift := c.computeDrift()
 	if interval+drift > 0 {
 		interval += drift
 	}
-	return time.Now().After(c.lastRan.Add(interval))
+	return c.clock.Now().After(c.lastRan.Add(interval))
 }
 
 func (c *component) UpdateLastRan() error {
-	_, err := psql.Update("components").
+	var lastRan time.Time
+	err := psql.Update("components").
 		Set("last_ran", sq.Expr("now()")).
 		Where(sq.Eq{
 			"id": c.id,
 		}).
+		Suffix("RETURNING last_ran").
 		RunWith(c.conn).
-		Exec()
+		QueryRow().Scan(&lastRan)
 	if err != nil {
 		return err
 	}
-
+	c.lastRan = lastRan
 	return nil
 }
 
+const maxDrift = 5 * time.Second
+
+// ComputeDrift computes a drift for components scheduler, the drift should help
+// distribute workloads more evenly across ATCs. When numGoroutineThreshold
+// is not set, drift will be a random value in range of [-1, 1] second. When
```

Comment:
just trying to understand why it drift differently when num Goroutine set/unset.

Reply:
For random drift, all ATCs are in the equal places, thus it doesn't make sense to add a 5-second drift. For numGoroutine based drift, when a ATC has 20K goroutines, obviously it should get bigger drift than a ATC that has only 0.1K goroutines.